### ISSUE TITLE
Dumping several database with async causes several issues

### ISSUE THREAD

**Issue by BuzzMyVideosDeployService** [bradzacher/mysqldump#21]
I'm trying to create dumps of several database, for some database i've selected tables and for others I've all the tables.

I used an async.forEach to run in parallel all the various database's dump but everytime I see that the module endup mixing up the tables with the database.

For example if the database 'AAA' have the tables '111' and '222' and the database 'BBB' have the table '333' and '444' I end up seeing errors like this:

(this is an actual error, just changed the tables and database name)

```
2016-11-09T16:03:31.418Z] SHOW CREATE TABLE `111` Error: ER_NO_SUCH_TABLE: Table 'BBB.111' doesn't exist
/Users/thecrius/Sites/db-sync/node_modules/mysql/lib/protocol/Parser.js:78
        throw err; // Rethrow non-MySQL errors
        ^

TypeError: Cannot read property '0' of undefined
    at /Users/thecrius/Sites/db-sync/node_modules/mysqldump/index.js:126:21
    at /Users/thecrius/Sites/db-sync/node_modules/mysqldump/node_modules/async/lib/async.js:726:13
    at /Users/thecrius/Sites/db-sync/node_modules/mysqldump/node_modules/async/lib/async.js:52:16
    at done (/Users/thecrius/Sites/db-sync/node_modules/mysqldump/node_modules/async/lib/async.js:241:17)
    at /Users/thecrius/Sites/db-sync/node_modules/mysqldump/node_modules/async/lib/async.js:44:16
    at /Users/thecrius/Sites/db-sync/node_modules/mysqldump/node_modules/async/lib/async.js:723:17
    at /Users/thecrius/Sites/db-sync/node_modules/mysqldump/node_modules/async/lib/async.js:167:37
    at Query._callback (/Users/thecrius/Sites/db-sync/node_modules/mq-node/dist/index.js:98:29)
    at Query.Sequence.end (/Users/thecrius/Sites/db-sync/node_modules/mysql/lib/protocol/sequences/Sequence.js:86:24)
    at Query.ErrorPacket (/Users/thecrius/Sites/db-sync/node_modules/mysql/lib/protocol/sequences/Query.js:94:8)
```

This happens because the table 111 is a table of the database AAA and not BBB.
Changing it to an async.forEachSeries fixed the problem (I suppose) but I ended up with this other error:

```
/Users/thecrius/Sites/db-sync/node_modules/mysqldump/index.js:163
                        callback(null,results.createSchemaDump.concat(results.createDataDump).join("\n\n"));
                                                                                              ^

RangeError: Invalid string length
    at Array.join (native)
    at Array.async.auto.getDataDump (/Users/thecrius/Sites/db-sync/node_modules/mysqldump/index.js:163:74)
    at listener (/Users/thecrius/Sites/db-sync/node_modules/mysqldump/node_modules/async/lib/async.js:605:42)
    at /Users/thecrius/Sites/db-sync/node_modules/mysqldump/node_modules/async/lib/async.js:544:17
    at _arrayEach (/Users/thecrius/Sites/db-sync/node_modules/mysqldump/node_modules/async/lib/async.js:85:13)
    at Immediate.taskComplete (/Users/thecrius/Sites/db-sync/node_modules/mysqldump/node_modules/async/lib/async.js:543:13)
    at runCallback (timers.js:637:20)
    at tryOnImmediate (timers.js:610:5)
    at processImmediate [as _immediateCallback] (timers.js:582:5)
```

No idea why this time because I hadn't info on which table was causing the issue.
The function is this one:

```
getDataDump:['createSchemaDump','createDataDump',function(callback,results){
			if(!results.createSchemaDump || !results.createSchemaDump.length) results.createSchemaDump=[];
			if(!results.createDataDump || !results.createDataDump.length) results.createDataDump=[];
			callback(null,results.createSchemaDump.concat(results.createDataDump).join("\n\n"));
		}
```

Maybe something with empty table?

I tried adding a console.log to the beginning of the function to see the content of `results` but that caused the script to end up with an out of memory:

```
FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - JavaScript heap out of memory
 1: node::Abort() [/usr/local/bin/node]
 2: node::FatalException(v8::Isolate*, v8::Local<v8::Value>, v8::Local<v8::Message>) [/usr/local/bin/node]
 3: v8::internal::V8::FatalProcessOutOfMemory(char const*, bool) [/usr/local/bin/node]
 4: v8::internal::Factory::NewRawTwoByteString(int, v8::internal::PretenureFlag) [/usr/local/bin/node]
 5: v8::internal::Object* v8::internal::StringReplaceGlobalAtomRegExpWithString<v8::internal::SeqTwoByteString>(v8::internal::Isolate*, v8::internal::Handle<v8::interna
l::String>, v8::internal::Handle<v8::internal::JSRegExp>, v8::internal::Handle<v8::internal::String>, v8::internal::Handle<v8::internal::JSArray>) [/usr/local/bin/node]
 6: v8::internal::Runtime_StringReplaceGlobalRegExpWithString(int, v8::internal::Object**, v8::internal::Isolate*) [/usr/local/bin/node]
 7: 0x35cdc4a092a7
```

I'll have to find another solution for my dumps but I'd gladly see this module improve as it's seems very clean and simple while still having some issue to be taken care of.

**Comment by BuzzMyVideosDeployService** [comment]
aaaand I commented with the wrong account 😅

**Comment by justechn** [comment]
I am also experiencing the same problem. Did you ever find a solution?

**Comment by bradzacher** [comment]
Should be fixed by #44
The out of memory problem is tricky - be careful you aren't doing too many things at once, or else node will just consume it's allocated amount.

This can be mitigated now though by using `returnFromFunction: false` for the data options (which will cause mysqldump to push data straight to the file on disk, rather than collecting the entire file then writing)
